File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/modules/runanywhere-core-onnx/src/main/kotlin/com/runanywhere/sdk/core/onnx/ONNX.kt
package com.runanywhere.sdk.core.onnx

import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * ONNX Runtime module for embedding services.
 *
 * Provides text-embedding capabilities using ONNX Runtime with models like
 * all-MiniLM. Speech primitives (STT/TTS/VAD) are served by the sherpa module.
 *
 * This is a thin wrapper that calls C++ backend registration.
 * All business logic is handled by the C++ commons layer.
 *
 * ## Registration
 *
 * ```kotlin
 * import com.runanywhere.sdk.core.onnx.ONNX
 *
 * // Register the backend (suspend, called once during SDK bootstrap)
 * ONNX.register()
 * ```
 *
 * ## Usage
 *
 * Services are accessed through the main SDK APIs - the C++ backend handles
 * service creation and lifecycle internally:
 *
 * ```kotlin
 * // Embeddings via public API
 * val embedding = RunAnywhere.embed(text, modelId = "all-minilm-l6-v2")
 * ```
 *
 * Matches iOS ONNX.swift exactly.
 */
object ONNX {
    private val logger = SDKLogger.onnx

    // MARK: - Module Info

    /** Current version of the ONNX Runtime module */
    const val version = "2.0.0"

    /** ONNX Runtime library version (underlying C library) */
    const val onnxRuntimeVersion = "1.24.3"

    /** Human-readable module name (ONNX). */
    const val moduleName: String = "ONNX"

    // MARK: - Registration State

    @Volatile
    private var isRegistered = false

    @Volatile
    private var isSherpaRegistered = false
    private val registrationMutex = Mutex()
    private val registrationLock = Any()

    // MARK: - Registration

    /**
     * Register ONNX backend with the C++ service registry.
     *
     * Calls `rac_backend_onnx_register()` to register the ONNX embedding
     * provider with the C++ commons layer. Suspend so that callers can
     * await module bootstrap from a coroutine scope.
     */
    suspend fun register() {
        registrationMutex.withLock {
            registerInternal()
        }
    }

    /**
     * Unregister the ONNX backend from C++ registry.
     */
    suspend fun unregister() {
        registrationMutex.withLock {
            if (!isRegistered) return

            if (isSherpaRegistered) {
                unregisterSherpaNative()
                isSherpaRegistered = false
                logger.info("Sherpa backend unregistered")
            }

            if (isRegistered) {
                unregisterNative()
                isRegistered = false
                logger.info("ONNX backend unregistered")
            }
        }
    }

    private fun registerInternal() {
        if (isRegistered) {
            logger.debug("ONNX already registered, returning")
            return
        }

        logger.info("Registering ONNX backend with C++ registry...")

        val result = registerNative()

        // Success or already registered is OK
        if (result != 0 && result != -4) { // RAC_ERROR_MODULE_ALREADY_REGISTERED = -4
            logger.error("ONNX registration failed with code: $result")
            // Don't throw - registration failure shouldn't crash the app
            return
        }

        isRegistered = true
        logger.info("ONNX backend registered successfully (embeddings)")
    }

    internal fun markSherpaRegistered() {
        isSherpaRegistered = true
    }

    // `canHandleSTT` / `canHandleTTS` / `canHandleVAD` deleted per
    // gaps/kotlin.md — mirrors SWIFT-DUP-CANHANDLE. The C++ plugin router
    // (`rac_router_*` / `rac_plugin_route`) is the only routing authority;
    // Kotlin-side substring matching was never called from the dispatch path.

    // MARK: - Auto-Registration

    /**
     * Enable auto-registration for this module.
     * Access this property to trigger C++ backend registration.
     */
    val autoRegister: Unit by lazy {
        synchronized(registrationLock) {
            registerInternal()
        }
    }
}

private val logger = SDKLogger.onnx

/**
 * JVM/Android implementation of ONNX native registration.
 *
 * Uses the self-contained ONNXBridge to register the backend,
 * mirroring the Swift ONNXBackend XCFramework architecture.
 *
 * The ONNX module has its own JNI library (librac_backend_onnx_jni.so)
 * that provides backend registration, separate from the main commons JNI.
 */
internal fun ONNX.registerNative(): Int {
    logger.debug("Ensuring commons JNI is loaded for service registry")
    // Ensure commons JNI is loaded first (provides service registry)
    RunAnywhereBridge.ensureNativeLibraryLoaded()

    logger.debug("Loading ONNX JNI library")
    // Load and use the dedicated ONNX JNI
    if (!ONNXBridge.ensureNativeLibraryLoaded()) {
        logger.error("Failed to load ONNX native library")
        throw UnsatisfiedLinkError("Failed to load ONNX native library")
    }

    logger.debug("Calling native ONNX register")
    val result = ONNXBridge.nativeRegister()
    logger.debug("Native ONNX register returned: $result")

    if (ONNXBridge.isSherpaLoaded) {
        val sherpaResult = ONNXBridge.nativeRegisterSherpa()
        if (sherpaResult == 0 || sherpaResult == -4) {
            ONNX.markSherpaRegistered()
            logger.info("Sherpa backend registered successfully")
        } else {
            logger.warning("Sherpa registration returned code: $sherpaResult")
        }
    } else {
        logger.info("Sherpa backend library not packaged; continuing with ONNX backend only")
    }
    return result
}

/**
 * JVM/Android implementation of ONNX native unregistration.
 */
internal fun ONNX.unregisterNative(): Int {
    logger.debug("Calling native ONNX unregister")
    val result = ONNXBridge.nativeUnregister()
    logger.debug("Native ONNX unregister returned: $result")
    return result
}

internal fun ONNX.unregisterSherpaNative(): Int {
    logger.debug("Calling native Sherpa unregister")
    val result = ONNXBridge.nativeUnregisterSherpa()
    logger.debug("Native Sherpa unregister returned: $result")
    return result
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/onnx/src/ONNX.ts
/**
 * @runanywhere/onnx - ONNX Runtime Module
 *
 * ONNX Runtime module wrapper for RunAnywhere React Native SDK.
 * Provides public API for module registration only.
 *
 * Model registration is done via RunAnywhere.registerModel() / RunAnywhere.registerMultiFileModel()
 * on the core SDK, matching the Swift SDK pattern where ONNX only exposes
 * register() and unregister().
 *
 * Reference: sdk/runanywhere-swift/Sources/ONNXRuntime/ONNX.swift
 */

import { ONNXProvider } from './ONNXProvider';
import { SDKLogger } from '@runanywhere/core/internal';

const logger = new SDKLogger('ONNX');

/**
 * ONNX Runtime Module
 *
 * Matches iOS: public enum ONNX: RunAnywhereModule
 *
 * Only provides backend registration. Model registration is done via
 * RunAnywhere.registerModel(framework: InferenceFramework.INFERENCE_FRAMEWORK_ONNX, ...) on the core SDK.
 *
 * ## Usage
 *
 * ```typescript
 * import { ONNX } from '@runanywhere/onnx';
 * import { RunAnywhere } from '@runanywhere/core';
 * import { ModelCategory, InferenceFramework } from '@runanywhere/proto-ts/model_types';
 *
 * // Register ONNX backend providers
 * await ONNX.register();
 *
 * // Register models via RunAnywhere (matching iOS pattern)
 * await RunAnywhere.registerModel({
 *   id: 'sherpa-onnx-whisper-tiny.en',
 *   name: 'Sherpa Whisper Tiny (ONNX)',
 *   url: '...',
 *   framework: InferenceFramework.INFERENCE_FRAMEWORK_ONNX,
 *   modality: ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION,
 *   memoryRequirement: 75_000_000
 * });
 * ```
 */
export const ONNX = {
  // Module-metadata constants (moduleId / moduleName / inferenceFramework /
  // capabilities / defaultPriority) were removed: the Swift source of truth
  // (Sources/ONNXRuntime/ONNX.swift) no longer declares them and nothing
  // consumed them here.

  /**
   * Register ONNX module with the SDK
   *
   * Registers both ONNX STT and TTS providers with ServiceRegistry,
   * enabling them to handle Sherpa-ONNX and Piper models.
   *
   * Matches iOS: static func register(priority: Int = 100)
   */
  async register(): Promise<boolean> {
    logger.info('Registering ONNX module (STT + TTS + VAD)');
    const registered = await ONNXProvider.register();
    if (registered) {
      logger.info('ONNX module registered');
    }
    return registered;
  },

  /**
   * Unregister ONNX module from the SDK
   *
   * Matches iOS: static func unregister()
   */
  async unregister(): Promise<boolean> {
    logger.info('Unregistering ONNX module');
    return ONNXProvider.unregister();
  },

  /**
   * Check if this module is registered with the native backend registry.
   */
  async isRegistered(): Promise<boolean> {
    return ONNXProvider.isRegistered();
  },
};
